### PR TITLE
Introduce x509 DID method with PKI validation.

### PR DESCRIPTION
**Pull request by rolandgroen**
Added support for x509 DID method including a new resolver and PKI validation. Adjusted metadata handling in key resolution across the codebase to include x509 specific fields and validation logic.

- Added support for did:x509 as a new DID resolver.
- Enabled validation of did:x509 VCs and VPs.
- Added corresponding tests.

### CODE REVIEW

File: vcr/verifier/signature_verifier.go
Comment (reinkrul) [inline vcr/verifier/signature_verifier.go:134]: why not always do this, regardless of the DID method? Probably less complex.

Although: you could decide to only do this when it's did:x509, then you can (more) safely return an error when `ParseString()` fails. Since the x5c chain is required for the process, I think it should fail if it can't parse the JWS? (or maybe even when x5c isn't there?)

File: vcr/verifier/verifier.go
Comment (reinkrul) [inline vcr/verifier/verifier.go:177]: duplicate, could use a util func, maybe in the x509 package?

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:76]: are these hashing algorithms fixed?

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:105]: I think this should've been subject malformed?

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:131]: this allows unsupported/unknown RDNs in the subject DN. There should be a default case that rejects the did:x509 (since it's not something checked against the certificate).

Comment (reinkrul) [reply vdr/didx509/resolver.go:131]: Also, it's missing `ST` and `STREET` (https://trustoverip.github.io/tswg-did-x509-method-specification/#json-data-model-for-x509-certificate-chains)

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:183]: The ID of the verification method needs to be the DID (you ahve this), appended with the `#0` fragment. I think you can set `Fragment` on the first argument to `0` to achieve this.

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:198]: missing `AddAuthenticationMethod`

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:211]: missing test coverage

File: vdr/didx509/resolver_test.go
Comment (reinkrul) [inline vdr/didx509/resolver_test.go:93]: however funny, better use non-existing/example domain (`example.com`)
```suggestion
		rootDID := did.MustParseDID("did:kaas:example.com:jonge")
```

File: vdr/didx509/resolver_test.go
Comment (reinkrul) [inline vdr/didx509/resolver_test.go:58]: duplicate test name

File: vdr/didx509/resolver_test.go
Comment (reinkrul) [inline vdr/didx509/resolver_test.go:49]: "2" isn't really informative

File: vdr/didx509/x509_utils.go
Comment (reinkrul) [inline vdr/didx509/x509_utils.go:69]: SAN -> San (Golang convention)

File: vdr/didx509/x509_utils.go
Comment (reinkrul) [inline vdr/didx509/x509_utils.go:154]:
```suggestion
// Supported algorithms are: "sha1", "sha256", "sha384", "sha512".
```

File: vcr/verifier/signature_verifier.go
Comment (rolandgroen) [reply vcr/verifier/signature_verifier.go:134]: It does not make a lot of sense to do this whenever the method is not x509, but somehow understandable if you would fill the headers anyway. There MIGHT be a situation where the fields are used outside the x509 spec..

About failing: 
I think it will fail later on when X509CertThumbprint et. al. is not filled and the method is x509.

Comment (rolandgroen) [reply vcr/verifier/signature_verifier.go:134]: Maybe the cleanest way would be to just stuff in the headers into the metadata. That would be way more generic.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: Yes.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: Correct. Fixed.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go:188]: First thing turned up at testing.

Comment (rolandgroen) [reply vdr/didx509/resolver.go:201]: The STREET and ST are added.

File: vdr/didx509/resolver_test.go
Comment (rolandgroen) [reply vdr/didx509/resolver_test.go]: Updated most domains and emails to example.com

File: vdr/didx509/resolver_test.go
Comment (rolandgroen) [reply vdr/didx509/resolver_test.go]: Resolved

File: vdr/didx509/x509_utils.go
Comment (rolandgroen) [reply vdr/didx509/x509_utils.go]: Changed.

File: vdr/didx509/resolver_test.go
Comment (rolandgroen) [reply vdr/didx509/resolver_test.go]: Updated all test names to more semantic names

File: vdr/didx509/x509_test_utils.go
Comment (gerardsn) [inline vdr/didx509/x509_test_utils.go:1]: filename should end in `_test.go` to prevent it from being used outside of tests

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: Check done

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: Set the fragment, added a check to the tests.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go:148]: Added tests, and PolicyNone. The latter will get removes as soon as we support multiple policies.

File: vdr/didx509/x509_test_utils.go
Comment (rolandgroen) [reply vdr/didx509/x509_test_utils.go:1]: Renamed the file.

File: vdr/didx509/jwt_utils.go
Comment (gerardsn) [inline vdr/didx509/jwt_utils.go:14]: inverting the conditions / removing all nested statements makes this a lot easier to read / maintain

File: vdr/didx509/resolver.go
Comment (gerardsn) [reply vdr/didx509/resolver.go:81]: Discussed with @stevenvegt: No. this is not acceptable. The `pkiValidator` must use hardfail: return an error when it has no CRL cache for the chain, or when the cache is stale, i.e. older than `pki.maxupdatefailhours`. For other/existing use-cases the failure mode should still be configurable.

File: vdr/didx509/resolver_test.go
Comment (gerardsn) [inline vdr/didx509/resolver_test.go:41]: FYI: Normally you would use `assert.ErrorIs(...)` to check for errors. `ErrorsIs` also works if the error is wrapped by another error

File: vdr/didx509/resolver_test.go
Comment (gerardsn) [inline vdr/didx509/resolver_test.go:68]: FYI: mocks always return the default value if none is specified. So you don't need to add `.Return(nil)`, only if you actually want to return an error.

File: vdr/didx509/x509_utils.go
Comment (gerardsn) [inline vdr/didx509/x509_utils.go:1]: seems to be missing a bunch of non-happy flow test cases

File: vdr/resolver/did.go
Comment (gerardsn) [inline vdr/resolver/did.go:1]: untested

**Review by gerardsn — CHANGES_REQUESTED** [review]
Very limited review since there is no reference to the spec + version being implemented.

Some general comments:
- files in `vdr/didx509` need more comments. I currently have no information on choices that were made to maintain this in the future.
- seems to be missing a bunch of non-happy flow tests, but the code coverage tool is also marking things untested that are tested... The truth is probably somewhere in between.

File: vcr/verifier/signature_verifier.go
Comment (rolandgroen) [reply vcr/verifier/signature_verifier.go:129]: Not completely sure I mocked the package function the right way, but a test is added.

File: vcr/verifier/verifier.go
Comment (rolandgroen) [reply vcr/verifier/verifier.go:171]: Added tests

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go:20]: Introduced a check for message.Signatures() == 1 and included tests for the case.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go:117]: Added a check, the nil is possible.

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go:18]: The conditions are removed, let the parse method figure it out.

File: vdr/didx509/resolver_test.go
Comment (rolandgroen) [reply vdr/didx509/resolver_test.go:41]: I did refactor the assertions, most to assert.ErrorIs, some didn't work that way. These are refactored to `assert.EqualError(t, err, str)`

File: vdr/didx509/resolver_test.go
Comment (rolandgroen) [reply vdr/didx509/resolver_test.go:68]: Removed the .Return(nil)

File: vdr/didx509/resolver.go
Comment (reinkrul) [reply vdr/didx509/resolver.go:83]: @gerardsn @stevenvegt why is this different from the other cases where this is used (why hard-fail here?).

I find it dangerous to choose to hard-fail here, since it causes data exchanges to be blocked when the issuer of the UZI cert backing the credential's CRL is unreachable or expired. It's already quite a long chain, with more and more failure modes (we also have the revocation status of the credential itself to check), so hard-fail makes it less robust.
That's why we implemented soft-fail in the first place, so how is it different here?

Comment (gerardsn) [reply vdr/didx509/resolver.go:83]: - before it was 'just' one of the layers of security. (Still requires did:nuts signatures to do anything)
- here it is the root of trust for your credentials. A CA should always be online, and even if it is offline for a few minutes, the cache is still available. Using softfail here means you accept credentials without checking the trust

I can make a PR that will make the required methods available

**Comment by gerardsn** [comment]
This is missing an e2e-test.

Certificate revocation is currently not checked. The `pki.Validator` can only check CRLs for CAs certs in the configured truststore. Validation fails if the CA is not in the truststore, but this specific error is ignored when configured for `softfail` (default).

An e2e-test should show 
- warning logs containing `certificate validation failed because CRL cannot be updated`
- error logs containing `Certificate CRL check softfail bypass. Might be unsafe, find cause of failure!`

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go:83]: @gerardsn I get what the general idea is, I just could use some help on how to implement this correctly. Should I create a new first class citizen for the pkiValidator instance with configuation elements, new defaults etc.? Or should I create my own instance with a specific configuration? Or should I clone / copy the TLS instance and just set the softfail = false?

Comment (gerardsn) [reply vdr/didx509/resolver.go:83]: I will create a PR later today that you can use here.

Comment (gerardsn) [reply vdr/didx509/resolver.go:83]: #3531 should give you `ValidateStrict` to use here

Comment (rolandgroen) [reply vdr/didx509/resolver.go:83]: What is procedure from here? Move on the this PR and add it later or wait for the https://github.com/nuts-foundation/nuts-node/pull/3531 to merge into main/master?

In the mean time I'm working on the latest issues.

Comment (gerardsn) [reply vdr/didx509/resolver.go:83]: PR is merged so yo can use `ValidateStrict` now. I believe @reinkrul was still going to compare this with the spec, so maybe resolve all conflicts in the meantime so the tests can run again.

File: vdr/didx509/x509_utils.go
Comment (rolandgroen) [reply vdr/didx509/x509_utils.go:1]: Added tests 😅

File: vdr/resolver/did.go
Comment (rolandgroen) [reply vdr/resolver/did.go:1]: Added tests.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: The master + PR is merged into this PR, code is changed to `ValidateStrict`.

**Review by reinkrul — APPROVED** [review]

File: vdr/didx509/jwt_utils.go
Comment (woutslakhorst) [inline vdr/didx509/jwt_utils.go:17]: should be private otherwise it's in the wrong place

File: vdr/didx509/jwt_utils.go
Comment (woutslakhorst) [inline vdr/didx509/jwt_utils.go:1]: missing copyrights

File: vdr/didx509/resolver.go
Comment (woutslakhorst) [inline vdr/didx509/resolver.go:22]: Missing godoc

File: vdr/didx509/resolver.go
Comment (woutslakhorst) [inline vdr/didx509/resolver.go:15]: missing godoc

File: vdr/didx509/resolver.go
Comment (woutslakhorst) [inline vdr/didx509/resolver.go:50]: godoc on public structs

Comment (woutslakhorst) [reply vdr/didx509/resolver.go:50]: and missing reference to w3c spec

File: vdr/didx509/validation.go
Comment (woutslakhorst) [inline vdr/didx509/validation.go:182]: public or private?

File: vdr/resolver/did.go
Comment (woutslakhorst) [inline vdr/resolver/did.go:196]: godoc

File: vdr/resolver/did.go
Comment (woutslakhorst) [inline vdr/resolver/did.go:208]: godoc

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
- refer to which spec is implemented in godoc comment
- copyrights
- godoc on public vars/functions/structs/methods

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go:17]: It's not private, so where should it go? `crypto/jwx.go`?

File: vdr/didx509/jwt_utils.go
Comment (gerardsn) [reply vdr/didx509/jwt_utils.go:1]: run `make fix-copyright`

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go:36]: This method is made private.

File: vdr/didx509/validation.go
Comment (rolandgroen) [reply vdr/didx509/validation.go]: Made private.

File: vdr/resolver/did.go
Comment (rolandgroen) [reply vdr/resolver/did.go:213]: Added doc.

File: vdr/resolver/did.go
Comment (rolandgroen) [reply vdr/resolver/did.go:199]: Added doc.

File: vdr/didx509/resolver.go
Comment (rolandgroen) [reply vdr/didx509/resolver.go]: Added docs

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go:19]: Added manually.

File: vdr/didx509/jwt_utils.go
Comment (rolandgroen) [reply vdr/didx509/jwt_utils.go]: Done.

**Review by woutslakhorst — APPROVED** [review]

**Review by gerardsn — APPROVED** [review]
Last comment: add did:x509 to the supported DID methods in `docs/pages/integrating/supported-protocols-formats.rst`

File: vdr/didx509/resolver.go
Comment (reinkrul) [inline vdr/didx509/resolver.go:215]: Does this work with percent-encoded values in the `did:x509`? The spec (https://trustoverip.github.io/tswg-did-x509-method-specification/#san-policy) states:

> ```
policy-name     = "san"
policy-value    = san-type ":" san-value
san-type        = "email" / "dns" / "uri"
san-value       = 1*idchar
```
> san-type is the SAN type and must be one of email, dns, or uri. Note that dn is not supported.
> san-value is the SAN value, percent-encoded.

**Review by stevenvegt — CHANGES_REQUESTED** [review]
Wait for it....
Let me try it out before merging.

File: vdr/didx509/x509_utils.go
Comment (rolandgroen) [reply vdr/didx509/x509_utils.go:107]: Well, the last one, actually. I'll take a look at this.

**Comment by rolandgroen** [comment]
The file `docs/pages/integrating/supported-protocols-formats.rst` file has been modified.

**Comment by rolandgroen** [comment]
The command `go fmt vdr/vdr.go` has been applied.

**Review by stevenvegt — APPROVED** [review]
I don't want to hold this one up longer than needed. We can finetune the logic in a later phase, since this feature is still very experimental.

**Comment by rolandgroen** [comment]
I did some end-to-end testing and some issues where found & fixed:
 *  The #0 fragment was missing in the generated vc, fixed that in the issuer code.
 *  The issuer added the permanentIdentifier by default, that does not work with the current NUTS code, so that feature has been hidden behind a flag in the issuer.
 * The x509 otheName extraction code did not check for other types of otheName values and crashed on the alternative host name. That has been fixed + tests have been added.

**Pull request merged**
